flow360/component/compress_upload.py

```diff
@@ -0,0 +1,102 @@
+"""
+Parallel Compress and Multiupload Flow360Resource to S3
+"""
+import bz2
+import concurrent.futures
+import os
+
+from rich.progress import (
+    BarColumn,
+    Progress,
+    TaskProgressColumn,
+    TextColumn,
+    TimeRemainingColumn,
+    TransferSpeedColumn,
+)
+
+from flow360.component.resource_base import Flow360Resource
+
+
+# pylint: disable=too-many-arguments, too-many-locals
+def compress_and_upload_chunks(
+    file_name: str,
+    upload_id: str,
+    remote_resource: Flow360Resource,
+    remote_file_name: str,
+    max_workers: int = 50,
+    chunk_length: int = 25 * 1024 * 1024,
+):
+    """
+    Compresses and uploads file chunks to a remote resource using Bzip2 compression.
+
+    Args:
+        file_name (str): The path to the input file that needs to be compressed
+        and uploaded.
+        upload_id (str): The ID of the multipart upload for the remote resource.
+        remote_resource (Flow360Resource): The remote resource to which the chunks
+        will be uploaded.
+        remote_file_name (str): The name of the remote file on the remote resource.
+        max_workers (int, optional): The maximum number of concurrent workers for
+        the thread pool (default is 50).
+        chunk_length (int, optional): The size (in bytes) of each chunk to be
+        compressed and uploaded (default is 25 MB).
+
+    Raises:
+        AssertionError: If the input 'file_name' does not exist or is not a regular file.
+    """
+
+    def upload_and_update(part_number, chunk):
+        future = executor.submit(
+            remote_resource.upload_part,
+            remote_file_name,
+            upload_id,
+            part_number,
+            chunk,
+        )
+        # Wait for the upload to finish and update the progress
+        result = future.result()
+        progress.update(task_id1, advance=len(chunk))
+        return result
+
+    assert os.path.isfile(file_name)
+    uploaded_parts = []  # Initialize an empty list to store the uploaded parts
+    futures = []
+    min_upload_size = 5 * 1024 * 1024
+    compressed_bytes = 0
+    with Progress(
+        TextColumn("[text.bold.green]{task.description}"),
+        BarColumn(),
+        TaskProgressColumn(),
+        TransferSpeedColumn(),
+        TimeRemainingColumn(),
+    ) as progress:
+        task_id = progress.add_task("[cyan]Compressing...", total=os.path.getsize(file_name))
+        # Rough estimate of size of compressed file
+        task_id1 = progress.add_task("Uploading...", total=os.path.getsize(file_name) * 0.37)
```

Comment:
what if the sum of chunks will be less then the estimated? Then the progress can stay at 99% which will look like unfinished. Can we fake the last upload to fill up to 0.37 * total?

Reply:
At the end of compression I update the total of this progress bar to accurately reflect how many compressed bytes there are, so this is not a problem